### PR TITLE
Add preamble, convert to bash, and remove env

### PR DESCRIPTION
**Pull request by WalterKolczynski-NOAA**
**Description**

Add preamble, convert to bash, and remove env

This is the first in a wave of commits to improve and standardize the scripts within global workflow. In this commit, all scripts run during execution are converted to bash and a preamble is added to every script that is not sourced by another script.

Every script executed during a forecast cycle is converted to bash. This was mostly straightforward, though there were a couple Korne-shell conventions (primarily using `typeset` to format strings) that had to be replaced with bash-compatable alternatives like `printf`. This in turn required a few modification to prevent zero-padded numbers from being treated as octals (other may have been pre-existing bugs).

The preamble contains a number of feature to standardize code and improve script flow and debugging.

- First, it uses two variables, `$STRICT` and `$TRACE` to control the behavior of `set`. When `$STRICT` is `"YES"`, error on undefined variables (`set -u`) and exit on non-zero return (`set -e`) are turned on. When `$TRACE` is `"YES"`, command trace (`set -x`) is turned on and a useful string is set to `$PS4` that gives the name and line number of the script. Both `$STRICT` and `$TRACE` default to `"YES"`. They also set up commands, `$ERR_EXIT_ON` and `$TRACE_ON`, that will restore the setting of each in the event a script needs to temporarily turn them off.

- Second, the preamble sets up primative timing of scripts using Posix `date`.

- Third, it echos the script is beginning and at what time.

- Finally, it also establishes a postamble function and sets it to run as a trap of EXIT. The postamble will use the end time to calculate the run time of the script, then print that the script has ended at what time, how long has elapsed, and the exit code. By setting this up as a trap instead of just calling it at the end of the script, it ensures the postamble is called even if the script exits early because there is an error.

- In response to this standardization, parts of scripts that performed these preamble functions (announcing start/end, `set -x`, etc) have been deleted. For some scripts where temporarily turning off `-x` or `-e` is needed, they now use `$ERR_EXIT_ON` and `$TRACE_ON` to return to the correct state afterwards, instead of blindly turning the setting back on.

- Additionally, some modifications were needed to comply with `set -eu`. Mostly taking care of undefined variables, but also a couple instances where a non-zero return code had to be dealt with.

If users wish to use their own preamble script instead, the default script can be overriden by setting `$PREAMBLE_SCRIPT` before the run begins.

Instance where scripts would print the full list of environment variables have been removed. These can be spot added back in to debug as necessary. Alternatively, a future PR will add them back in in a standardized way.

`rstprod.sh` is added to the link list from gsi_monitor.fd, as it is needed for the radmon scripts.

The placeholders for AWIPS and GEMPAK in the Hera and Orion environment scripts were replaced with the correct definitions. There were also other modifications to AWIPS and GEMPAK scripts to get it working for development (AWIPS still isn't and will be fixed in the future).

GSI scripts that were brought in recently had all of their backticks replaced with `$(  )` as was done with all other script previously.

Refs: #397

**Type of change**
- [x] Script maintenance

**How Has This Been Tested?**
- [x] Cycled test on Hera with BUFRSND, GEMPAK, AWIPS, WAFS
- [x] Coupled test on Hera
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by WalterKolczynski-NOAA** [comment]
Cycled and coupled confirmed bitwise identical to develop on Hera.

AWIPS fails, but gets farther than develop. More effort will be needed to get that working.

File: env/HERA.env [hunk 1]
```diff
@@ -261,9 +261,17 @@ elif [ $step = "postsnd" ]; then
...
+    export APRUN_AWIPSCFP="$launcher -n $npe_awips --multi-prog"
 
 elif [ $step = "gempak" ]; then
 
-     echo "WARNING: $step is not enabled on $machine!"
+    nth_max=$(($npe_node_max / $npe_node_gempak))
+
+    export NTHREADS_GEMPAK=${nth_gempak:-1}
+    [[ $NTHREADS_GEMPAK -gt $nth_max ]] && export NTHREADS_GEMPAK=$nth_max
+    export APRUN="$launcher -n $npe_gempak --multi-prog"
```
Comment (aerorahul) [inline env/HERA.env:276]: Is `gempak` supported on Hera, Orion?

File: jobs/rocoto/anal.sh [hunk 2]
```diff
@@ -1,4 +1,9 @@
-#!/bin/bash -x
+#! /usr/bin/env bash
+
+PREAMBLE_SCRIPT="${PREAMBLE_SCRIPT:-$HOMEgfs/ush/preamble.sh}"
+if [ -f "${PREAMBLE_SCRIPT}" ]; then
+  source $PREAMBLE_SCRIPT
+fi
```
Comment (aerorahul) [inline jobs/rocoto/anal.sh]: What is the point of sourcing `PREAMPLE_SCRIPT` in `anal.sh` when it is sourced in the J-JOB.

I can see you need to source it in jobs that don't have a j-job e.g. `aerosol_init.sh` above.

File: jobs/rocoto/awips.sh [hunk 3]
```diff
@@ -73,7 +78,7 @@ for fhr in $fhrlst; do
     fhmax=84
     if [ $fhr -ge $fhmin -a $fhr -le $fhmax ] ; then
 	if [[ $(expr $fhr % 3) -eq 0 ]]; then
-            fhr3=$(printf %03i $fhr)
+            fhr3=$(printf %03d $((10#$fhr)))
```
Comment (aerorahul) [inline jobs/rocoto/awips.sh:78]: why is the one on the left wrong?

File: jobs/rocoto/earc.sh [hunk 4]
```diff
@@ -134,20 +139,26 @@ if [ $ENSGRP -eq 0 ]; then
...
+        ${ERR_EXIT_ON:-set -eu}
     fi
 
     #-- Archive online for verification and diagnostics
     [[ ! -d $ARCDIR ]] && mkdir -p $ARCDIR
     cd $ARCDIR
 
-    $NCP $ROTDIR/enkf${CDUMP}.$PDY/$cyc/$COMPONENT/${CDUMP}.t${cyc}z.enkfstat         enkfstat.${CDUMP}.$CDATE
-    $NCP $ROTDIR/enkf${CDUMP}.$PDY/$cyc/$COMPONENT/${CDUMP}.t${cyc}z.gsistat.ensmean  gsistat.${CDUMP}.${CDATE}.ensmean
+    if [[ -f $ROTDIR/enkf${CDUMP}.$PDY/$cyc/$COMPONENT/${CDUMP}.t${cyc}z.enkfstat ]]; then
```
Comment (aerorahul) [inline jobs/rocoto/earc.sh]: why not use `safe_copy` here?

File: jobs/rocoto/waveprep.sh [hunk 5]
```diff
@@ -1,4 +1,9 @@
-#!/bin/bash -x
+#! /usr/bin/env bash
+
+PREAMBLE_SCRIPT="${PREAMBLE_SCRIPT:-$HOMEgfs/ush/preamble.sh}"
+if [ -f "${PREAMBLE_SCRIPT}" ]; then
+  source $PREAMBLE_SCRIPT
+fi
```
Comment (aerorahul) [inline jobs/rocoto/waveprep.sh]: capping back at the last rocoto job wrapper, why is this necessary here, when the j-job does it first thing.

File: scripts/exgdas_atmos_chgres_forenkf.sh [hunk 6]
```diff
@@ -18,11 +18,9 @@
...
 ################################################################################
 
-#  Set environment.
-export VERBOSE=${VERBOSE:-"YES"}
-if [ $VERBOSE = "YES" ]; then
-   echo $(date) EXECUTING $0 $* >&2
-   set -x
+PREAMBLE_SCRIPT="${PREAMBLE_SCRIPT:-$HOMEgfs/ush/preamble.sh}"
+if [ -f "${PREAMBLE_SCRIPT}" ]; then
+  source $PREAMBLE_SCRIPT
```
Comment (aerorahul) [inline scripts/exgdas_atmos_chgres_forenkf.sh]: Why is this needed in the exscript when the JJob has already set the necessary tracing, error handling etc.
This seems like a lot of duplication.

File: scripts/exgfs_atmos_awips_20km_1p0deg.sh [hunk 7]
```diff
@@ -75,7 +77,7 @@ echo "#######################################"
 echo " Process GRIB AWIP GRIB2 PRODUCTS      "
 echo "#######################################"
 echo " "
-set -x
+${TRACE_ON:-set -x}
```
Comment (aerorahul) [inline scripts/exgfs_atmos_awips_20km_1p0deg.sh:77]: What is the use of the preamble if every script will set/reset this.

File: scripts/exgfs_wave_post_gridded_sbs.sh [hunk 8]
```diff
@@ -55,7 +55,7 @@
   echo "Starting at : $(date)"
   echo '-------------'
   echo ' '
-  [[ "$LOUD" = YES ]] && set -x
+  ${TRACE_ON:-set -x}
```
Comment (aerorahul) [inline scripts/exgfs_wave_post_gridded_sbs.sh:54]: Technically this is correct, but the entire `set +x` before and echo and `set -x` after the echo is useless when `TRACE_ON` is set.
What real difference does it make to do `set +x` for an echo statement when the default project wide is `set -x`?

**Review by aerorahul — CHANGES_REQUESTED** [review]
Several questions.  See inline comments.

File: env/HERA.env (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply env/HERA.env:276]: There is a gempak module available on both machines.

File: jobs/rocoto/awips.sh (hunk 3 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/awips.sh:78]: bash has an annoying tendency to treat zero-padded numbers as octals. The 10# forces it to be treated as a decimal (which is then zero-padded again; it's annoying). Without this is will complain about numbers like '018', because 8 is an invalid octal digit.

File: jobs/rocoto/earc.sh (hunk 4 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/earc.sh]: Probably just missed it. Will update. I think we need to have someone review this and identify which files we should fatally fail on and which we can continue on without.

File: jobs/rocoto/anal.sh (hunk 2 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/anal.sh]: Otherwise you don't have it while the rocoto pre-job runs, and there are sometimes issues (especially undefined variables) there too.

File: scripts/exgfs_atmos_awips_20km_1p0deg.sh (hunk 7 above)
Comment (WalterKolczynski-NOAA) [reply scripts/exgfs_atmos_awips_20km_1p0deg.sh:77]: This is only used if the script is using `set +x` before this point, usually to echo a bunch of lines (the wave scripts do a lot of this). This makes sure after that section has completed, it returns to its prior state instead of blinding turning `set -x` back on. This way if you turn `TRACE=NO`, then it won't turn the trace on here.

File: scripts/exgfs_wave_post_gridded_sbs.sh (hunk 8 above)
Comment (WalterKolczynski-NOAA) [reply scripts/exgfs_wave_post_gridded_sbs.sh:54]: I didn't want to go through all the (mostly wave) scripts and change this overall paradigm just yet (this PR is big enough). When I move on to giving these scripts closer attention, I plan on revisiting this. Simplest solution then will be to leave trace on (if it is on) and echo a heredoc instead of individual echo lines (which is how all multiline echos should be handled).

File: jobs/rocoto/awips.sh (hunk 3 above)
Comment (aerorahul) [reply jobs/rocoto/awips.sh:78]: What is the incoming value of `fhr`?  Does it have the `0` padded?
If it doesn't, you would not get the error.

File: jobs/rocoto/awips.sh [hunk 9]
```diff
@@ -1,4 +1,9 @@
-#!/bin/bash -x
+#! /usr/bin/env bash
```
Comment (aerorahul) [inline jobs/rocoto/awips.sh:1]: I also don't understand this job.
It doesn't have a typical j-job?
How is this being called in ops?

File: jobs/rocoto/awips.sh (hunk 3 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/awips.sh:78]: Probably not? Something to evaluate later.

File: jobs/rocoto/awips.sh (hunk 9 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/awips.sh:1]: The jjobs are coded in `AWIPS20SH` and `AWIPSG2SH`, which are set in [config.awips](https://github.com/NOAA-EMC/global-workflow/blob/develop/parm/config/config.awips). Those are then called by this script.

Yeah, this arrangement doesn't make sense to me either. Still really want to get rid of all these `jobs/rocoto/` scripts anyway and call j-jobs directly.

File: scripts/exgdas_atmos_chgres_forenkf.sh (hunk 6 above)
Comment (WalterKolczynski-NOAA) [reply scripts/exgdas_atmos_chgres_forenkf.sh]: The settings aren't handed down to daughter scripts (unless sourced, of course).

Comment (WalterKolczynski-NOAA) [reply scripts/exgdas_atmos_chgres_forenkf.sh]: Also, even if the `set` built-in were, the way I have the postamble set up with the trap wouldn't work.

File: jobs/rocoto/awips.sh (hunk 9 above)
Comment (aerorahul) [reply jobs/rocoto/awips.sh:1]: I see whats going with `awips.sh`.
This is being treated like `arch.sh` or any other without a j-job.
The j-job does not follow the convention of the other global-workflow j-jobs.
The configs are sourced in this rocoto wrapper job and probably not in the j-job.

File: jobs/rocoto/earc.sh (hunk 4 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/earc.sh]: Oh, it's not that I missed it. I fixed this bit; then later I realized it was going to keep being an issue and made a function instead, but never replaced the first edit with the function. Fixed now.

We could still use some feedback on which of these (if any) should be fatal. These changes are just ignoring missing files so it completes without regard to whether we _should_ or not.

Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/earc.sh]: Wait, safe_copy isn't even in this script, it's in the other archive script. Should I copy that function here, even though it is only needed a few times in this script?

Comment (aerorahul) [reply jobs/rocoto/earc.sh]: perhaps put that in a `ush/workflow_utils.sh` and source it.  You could call it `file_utils.sh` if you wish to use other file/directory manipulations.

File: jobs/rocoto/anal.sh [hunk 10]
```diff
@@ -10,4 +12,6 @@ status=$?
 # Execute the JJOB
 $HOMEgfs/jobs/JGLOBAL_ATMOS_ANALYSIS
 status=$?
-exit $status
+
+
+exit ${status}
```
Comment (KateFriedman-NOAA) [inline jobs/rocoto/anal.sh:17]: Why {} for $status here but not in all of the other jobs/rocoto scripts?

File: jobs/rocoto/vrfy.sh [hunk 11]
```diff
@@ -56,7 +58,7 @@ echo
 echo "=============== START TO GENERATE QUARTER DEGREE GRIB1 FILES ==============="
 if [ $MKPGB4PRCP = "YES" -a $CDUMP = "gfs" ]; then
     if [ ! -d $ARCDIR ]; then mkdir $ARCDIR ; fi
-    nthreads_env=${OMP_NUM_THREADS:-1} # get threads set in env
+    nthreads_env=${OMP_NUM_THREADS:-1} # get threads set in env
```
Comment (KateFriedman-NOAA) [inline jobs/rocoto/vrfy.sh]: Remove extra whitespace at end of line 61.

File: jobs/rocoto/earc.sh (hunk 4 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/earc.sh]: Done. I'm not going to go through every script (yet) and replace all the file checking with safe_copy(), but did apply it to earc.

Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/earc.sh]: Maybe I should rename it non_blocking_copy?

Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/earc.sh]: Changed to nb_copy

File: ush/preamble.sh [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+	end_time=$(date +%s)
+	elapsed_sec=$((end_time - start_time))
+	elapsed=$(date -d@${elapsed_sec} -u +%H:%M:%S)
+
+	# Announce the script has ended, then pass the error code up
+	echo "End ${script} at $(date -u) with error code ${rc:-0} (time elapsed: ${elapsed})"
+	exit ${rc}
+}
+
+# Place the postable in a trap so it is always called no matter how the script exits
```
Comment (KateFriedman-NOAA) [inline ush/preamble.sh]: Spelling mistake, missing the "m" in "postable"?

**Review by KateFriedman-NOAA — APPROVED** [review]
Left a few minor comments. Looks fantastic! Approve pending resolution of other reviewer comments and any remaining testing. Thank you for these updates @WalterKolczynski-NOAA !

File: jobs/rocoto/anal.sh (hunk 10 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/anal.sh:17]: This was the one I happened to change at some point for some reason. Eventually I will likely make sure `{ }` is always used.

File: ush/preamble.sh (hunk 12 above)
Comment (WalterKolczynski-NOAA) [reply ush/preamble.sh]: done

File: jobs/rocoto/vrfy.sh (hunk 11 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/vrfy.sh]: done

File: jobs/rocoto/anal.sh (hunk 10 above)
Comment (KateFriedman-NOAA) [reply jobs/rocoto/anal.sh:17]: Roger that!

**Review by aerorahul — APPROVED** [review]
Looks good. 
Hopefully now the stdout will look nice and make it easier to trace failures

File: jobs/rocoto/awips.sh (hunk 3 above)
Comment (aerorahul) [reply jobs/rocoto/awips.sh:78]: Ok.

File: jobs/rocoto/anal.sh (hunk 2 above)
Comment (aerorahul) [reply jobs/rocoto/anal.sh]: Something to look for to avoid repetition later.

File: scripts/exgdas_atmos_chgres_forenkf.sh (hunk 6 above)
Comment (aerorahul) [reply scripts/exgdas_atmos_chgres_forenkf.sh]: to be evaluated later.

File: scripts/exgfs_wave_post_gridded_sbs.sh (hunk 8 above)
Comment (aerorahul) [reply scripts/exgfs_wave_post_gridded_sbs.sh:54]: Understood.

**Comment by NeilBarton-NOAA** [comment]
Was this tested with DA? I'm getting unbound variable errors when running

**Comment by WalterKolczynski-NOAA** [comment]
> Was this tested with DA? I'm getting unbound variable errors when running

Yes, except the new JEDI-based stuff.
